selection.go

```diff
@@ -0,0 +1,267 @@
+		a.pendingBindingRequests = a.pendingBindingRequests[overflow:]
+	}
+
+	a.pendingBindingRequests = append(a.pendingBindingRequests, bindingRequest{
+		transactionID: transactionID,
+		destination:   remote.addr(),
+	})
+	a.sendSTUN(msg, local, remote)
+}
+
+type ControlledSelector struct {
+	agent      *Agent
+	nominationTxID []byte
+	validateNominationTxID []byte
+}
+
+func (s *ControlledSelector) Start() {}
+
+func (s *ControlledSelector) PingCandidate(local, remote *Candidate) {
+	a := s.agent
+	transactionID := stun.GenerateTransactionID()
+
+	msg, err := stun.Build(stun.ClassRequest, stun.MethodBinding, transactionID,
+		&stun.Username{Username: a.remoteUfrag + ":" + a.localUfrag},
+		&stun.IceControlled{TieBreaker: a.tieBreaker},
+		&stun.Priority{Priority: local.Priority()},
+		&stun.MessageIntegrity{
+			Key: []byte(a.remotePwd),
+		},
+		&stun.Fingerprint{},
+	)
+
+	if err != nil {
+		a.log.Error(err.Error())
+		return
+	}
+
+	a.log.Tracef("ping STUN from %s to %s\n", local.String(), remote.String())
+	if overflow := len(a.pendingBindingRequests) - maxPendingBindingRequests; overflow > 1 {
+		a.log.Debugf("Discarded %d pending binding requests, pendingBindingRequests is full", overflow)
+		a.pendingBindingRequests = a.pendingBindingRequests[overflow:]
+	}
+
+	a.pendingBindingRequests = append(a.pendingBindingRequests, bindingRequest{
+		transactionID: transactionID,
+		destination:   remote.addr(),
+	})
+	a.sendSTUN(msg, local, remote)
+}
+
+func (s *ControlledSelector) HandleSucessResponse(m *stun.Message, local, remote *Candidate, remoteAddr net.Addr) {
+	// TODO according to the standard we should specifically answer a failed nomination:
+	// https://tools.ietf.org/html/rfc8445#section-7.3.1.5
+	// If the controlled agent does not accept the request from the
+	// controlling agent, the controlled agent MUST reject the nomination
+	// request with an appropriate error code response (e.g., 400)
+	// [RFC5389].
+
+	ok, pendingRequest := s.agent.handleInboundBindingSuccess(m.TransactionID)
+	if !ok {
+		s.agent.log.Errorf("discard message from (%s), invalid TransactionID %s", remote, m.TransactionID)
+		return
+	}
+
+	transactionAddr := pendingRequest.destination
+
+	// Assert that NAT is not symmetric
+	// https://tools.ietf.org/html/rfc8445#section-7.2.5.2.1
+	if !addrEqual(transactionAddr, remoteAddr) {
+		s.agent.log.Debugf("discard message: transaction source and destination does not match expected(%s), actual(%s)", transactionAddr, remote)
+		return
+	}
+
+	s.agent.log.Tracef("inbound STUN (SuccessResponse) from %s to %s", remote.String(), local.String())
+	s.agent.addValidPair(local, remote)
+
+	if bytes.Equal(s.validateNominationTxID, m.TransactionID) {
+		s.agent.sendBindingSuccess(s.nominationTxID, local, remote)
+	}
+}
+
+func (s *ControlledSelector) HandleBindingRequest(m *stun.Message, local, remote *Candidate) {
+	_, useCandidate := m.GetOneAttribute(stun.AttrUseCandidate)
+
+	if useCandidate {
```

Comment:
Should we allow `useCandidate` for different pairs? 

Maybe some sort of assert/check so we don't end up in the state where local/remote have different pairs.